packages/kit/src/runtime/server/page/fetch.js

```diff
@@ -23,178 +22,144 @@ export function create_fetch({ event, options, state, route, prerender_default }
-			}
+						return new Response(options.read(file), {
+							headers: type ? { 'content-type': type } : {}
+						});
+					}
 
-			response = await respond(
-				new Request(new URL(requested, event.url).href, { ...opts }),
-				options,
-				{
-					prerender_default,
-					...state,
-					initiator: route
+					return await fetch(request);
 				}
-			);
 
-			if (state.prerendering) {
-				dependency = { response, body: null };
-				state.prerendering.dependencies.set(resolved, dependency);
-			}
-		} else {
-			// external
-			if (resolved.startsWith('//')) {
-				requested = event.url.protocol + requested;
-			}
+				if (request.credentials !== 'omit') {
+					const authorization = event.request.headers.get('authorization');
 
-			const url = new URL(requested);
-
-			// external fetch
-			// allow cookie passthrough for "same-origin"
-			// if SvelteKit is serving my.domain.com:
-			// -        domain.com WILL NOT receive cookies
-			// -     my.domain.com WILL receive cookies
-			// -    api.domain.dom WILL NOT receive cookies
-			// - sub.my.domain.com WILL receive cookies
-			// ports do not affect the resolution
-			// leading dot prevents mydomain.com matching domain.com
-			if (`.${url.hostname}`.endsWith(`.${event.url.hostname}`) && opts.credentials !== 'omit') {
-				const cookie = event.request.headers.get('cookie');
-				if (cookie) opts.headers.set('cookie', cookie);
-			}
+					// combine cookies from the initiating request with any that were
+					// added via set-cookie
+					const combined_cookies = { ...initial_cookies };
 
-			// we need to delete the connection header, as explained here:
-			// https://github.com/nodejs/undici/issues/1470#issuecomment-1140798467
-			// TODO this may be a case for being selective about which headers we let through
-			opts.headers.delete('connection');
+					for (const cookie of cookies) {
+						if (!domain_matches(event.url.hostname, cookie.domain)) continue;
+						if (!path_matches(url.pathname, cookie.path)) continue;
 
-			const external_request = new Request(requested, /** @type {RequestInit} */ (opts));
-			response = await options.hooks.externalFetch.call(null, external_request);
+						combined_cookies[cookie.name] = cookie.value;
+					}
 
-			if (opts.mode === 'no-cors') {
-				response = new Response('', {
-					status: response.status,
-					statusText: response.statusText,
-					headers: response.headers
-				});
-			} else {
-				if (url.origin !== event.url.origin) {
-					const acao = response.headers.get('access-control-allow-origin');
-					if (!acao || (acao !== event.url.origin && acao !== '*')) {
-						throw new Error(
-							`CORS error: ${
-								acao ? 'Incorrect' : 'No'
-							} 'Access-Control-Allow-Origin' header is present on the requested resource`
-						);
+					const cookie = Object.entries(combined_cookies)
+						.map(([name, value]) => `${name}=${value}`)
+						.join('; ');
+
+					if (cookie) {
+						request.headers.set('cookie', cookie);
```

Comment:
If the user sets a cookie in `handleFetch` and then calls the built-in `fetch` that's passed to it, will this clobber whatever cookie header they'd set?

Reply:
bugger. yes, it will. do we try and merge cookies, or if someone explicitly set `cookie` already do we just leave it be?

Reply:
After a bit more time reading the code: It looks like the `cookies` array is to make sure that we copy over the cookies from `set-cookie` in responses to the cookies we send in subsequent `fetch`es as part of this same `load` - is that correct? I'd missed that the first couple of times around, and there's probably a better name for that. (`new_cookies`? I don't know.)

I _think_ it still makes sense to automatically overlay the `set-cookie` cookies on whatever underlying cookies we had. If someone's `handleFetch` calls `fetch` with a `cookie` header, we just use that as `initial_cookies`, rather than an all-or-nothing of the `cookie` header from the original incoming request. Does that sound reasonable?

Looking at the code, I haven't quite worked out what would happen if someone decided to use `globalThis.fetch` in `handleFetch`. Would it just skip the `cookie` and `authorization` stuff, but still create the `Response` proxy so that if the user's `load` uses it, it gets serialized into the HTML? If that all still works, then that sounds like a reasonable escape hatch, and I feel more comfortable saying that the `handleFetch` `fetch` should overlay the `set-cookie` cookies unconditionally without any way to control that, and it would just overlay them on top of whatever `cookie` header was passed to `fetch` (if any), or else on top of whatever `cookie` header was present on the initial incoming request (if the domains match by our old logic), or else on top of nothing.

Reply:
> Would it just skip the cookie and authorization stuff, but still create the Response proxy so that if the user's load uses it, it gets serialized into the HTML?

Yep. How does https://github.com/sveltejs/kit/pull/6565/commits/f24ba0dc0fb9011648817617cffb58bd305ee4e6 look?

Reply:
I'm misunderstanding the code, or else I don't think this looks right. If `url.origin !== event.url.origin`, it doesn't look like this will respect the `new_cookies`.

I _think_ everything else looks right, but it's a bit hard to tell because a lot of this code looks like it's more complicated than it needs to be. I think it could be a bit more DRY, but that doesn't have to happen now, and I don't have specific recommendations of what to change.

In the case where there are both automatically inherited cookies and an explicitly specified `cookie` header, my suggestion above had been to completely replace the automatically inherited cookies with the explicitly specified ones, and I see that this will instead merge them, but I don't have a strong opinion on which would be less confusing. I don't know whether we need to distinguish (somehow) between `cookie` headers that came from `handleFetch` versus those that came from the `fetch` call in `load`. I hope not. How does the real browser `fetch` behave when you manually pass it a `cookie` header? What does that do to the cookies it would have automatically sent over?

Reply:
I _think_ I fixed that

> How does the real browser fetch behave when you manually pass it a cookie header? 

Nothing, it's a [forbidden header name](https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name). It's silently ignored

Reply:
Looks good to me!